python/whylogs/api/logger/multi_dataset_logger/multi_dataset_rolling_logger.py

```diff
@@ -0,0 +1,362 @@
+
+
+class DatasetProfileContainer:
+    """
+    A container that abstracts over different types of profiles.
+
+    This does the work of deciding how to track data and how to create profiles given a DatasetSchema.
+    This can only be used to manage a single entity for a given time. For example, this can represent
+    a normal DatasetProfile or segment that has a given dataset timestamp.
+    """
+
+    _target: Union[DatasetProfile, SegmentCache]
+
+    def __init__(self, dataset_timestamp: int, schema: Optional[DatasetSchema]) -> None:
+        self._schema: Optional[DatasetSchema] = schema
+        self._active = True
+        self._dataset_timestamp = datetime.fromtimestamp(dataset_timestamp / 1000.0, tz=tz.tzutc())
+        if self._has_segments() and schema is not None:  # Need the duplicate None check for type safety
+            self._target = SegmentCache(schema=schema)
+        else:
+            self._target = DatasetProfile(dataset_timestamp=self._dataset_timestamp)
+
+    def _has_segments(self) -> bool:
+        return self._schema is not None and self._schema.segments is not None
+
+    def _track_segments(self, data: TrackData) -> None:
+        if self._schema is None:
+            raise Exception("Schema missing in logger while using segments")
+
+        if not isinstance(self._target, SegmentCache):
+            raise Exception("Segment cache missing in logger while using segments")
+
+        if isinstance(data, List):
+            for row in data:
+                segment_processing(self._schema, row=row, segment_cache=self._target)
+        else:
+            segment_processing(self._schema, data, segment_cache=self._target)
+
+    def _track_profile(self, data: TrackData) -> None:
+        if not isinstance(self._target, DatasetProfile):
+            raise Exception("Dataset profile missing in logger")
+
+        if isinstance(data, List):
+            for row in data:
+                self._target.track(row=row)
+        else:
+            self._target.track(data)
+
+    def track(self, data: TrackData) -> None:
+        """
+        Track data against the contained profile or segment.
+        """
+        if not self._active:
+            # Should never happen
+            raise Exception("Profile container no longer active.")
+
+        if self._has_segments():
+            self._track_segments(data)
+        else:
+            self._track_profile(data)
+
+    def to_result_set(self) -> ResultSet:
+        """
+        Get the ResultSet of the contained profile/segment.
+
+        This doesn't have any side effects. It generates a ResultSet of whatever
+        is inside when this is called.
+        """
+        try:
+            if isinstance(self._target, SegmentCache):
+                return self._target.flush(dataset_timestamp=self._dataset_timestamp)
+            elif isinstance(self._target, DatasetProfile):
+                return ProfileResultSet(self._target)
```

Comment:
Is it possible this profile has processed 2 out of three columns of a row?

Reply:
You mean something like this?

```python

row1 = {'col1': 2}
row2 = {'col2': 3}
logger.log(row1, timestamp_ms=foo)
logger.log(row2, timestamp_ms=foo-one_hour)
```

Then there would be two profiles with different column layouts yeah. Is that bad?

Reply:
different layouts is fine. We can merge these correctly. I was wondering more about logging large dataframes that could take significant time and then calling flush while this processing is in flight, but maybe that is a question better scoped to multi-process changes.